### ISSUE TITLE
The signature verified correctly with the key contained in the signature, but that key is not trusted

### ISSUE THREAD

**Issue by abinjaik** [KentorIT/authservices#659]
I have my SAML2 endpoint in Azure and tried to integrate with Kentor Sample Mvc application. Everything went fine, Even I can see the SAMLRESPONSE with claims(attributes) in it. But I am getting a certificate error like "The signature verified correctly with the key contained in the signature, but that key is not trusted.' occurred in Kentor.AuthServices.dll" . Could anyone please help me on this?

What I could see is this error comes from XmlHelpers.cs file's _VerifySignature_ private function. LOC 293.

` if (containedKey != null && signedXml.CheckSignature(containedKey, true))
                {
                    throw new InvalidSignatureException("The signature verified correctly with the key contained in the signature, but that key is not trusted.");
                }`

Also came to realize that, signedXml.CheckSignature is returning false always. I dont have a signingcertificate for this Azure SAML2 endpoint ? What to do? or Where from Azure AD will I get the Cert?

**Comment by explunit** [comment]
There was some previous discussion of similar things in #497 and #493 

Do any of the suggestion there solve your issue?

**Comment by AndersAbel** [comment]
Azure changes rolls over (changes) the signing keys on a regular basis. Don't configure any keys by hand. Let AuthServices read the metadata instead.

**Comment by augustg** [comment]
I think I ran into the same issue trying to use azure as IdP.
The exception is being thrown from VerifySignature in XmlHelpers.cs.
The metadata (which is automatically downloaded) contains 3 signing certificates and one certificate within the signature itself. The latter cert is the exact same as the first of the 3 signing certificates.
In VerifySignature() the 3 signing certs are checked first and they all fail and the method falls through to checking against the certificate within the signature. Which partly succeeds, but ultimately fails because the certificate is not trusted (adding it to the root cert store in windows makes this check pass).

Another odd thing I noticed is that the asymmetric algorithm used when checking the 3 signing certs is hardcoded to SignedXml.XmlDsigRSASHA1Url. However checking the metadata reveals that the signature method used is "http://www.w3.org/2001/04/xmldsig-more#rsa-sha256".
I figured this was the problem so i just passed in "signedXml.Signature.SignedInfo.SignatureMethod" to GetAsymmetricAlgorithm but it did not solve the problem. Signature check still failed.
I dont have enough knowledge about saml to know how it's supposed to work and my outstanding questions are:
1. Is my reasoning above correct? Should we check the signature method and choose an algorithm based on that?
2. Why is it ok to NOT validate (if it's trusted etc) the certificate if we are checking against the 3 signing certs but not if we are checking against the cert contained in the signature itself. (ValidateCertificate takes a bool that is false but the later call to CheckSignature takes a bool that is hardcoded to true).
